Add support for serving Certificate Transparency (SCTs via TLS Extension)

aeijdenberg opened the pull request:
Add support for Certificate Transparency (http://www.certificate-transparency.org/) TLS Extension (https://tools.ietf.org/html/rfc6962#section-3.3.1).  This allows the TLS server to serve a list of SCTs returned by logs as part of the TLS handshake meaning that clients RFC6962 compliant clients that support (and eventually require) Certificate Transparency can validate that the certificate served has been written to one or more publicly auditable tamper-proof logs.

To demonstrate:

```
$ LD_LIBRARY_PATH=lib bin/s2nd localhost 4443
```

and in another terminal, verify that data is returned for TLS extension 18:

```
$ libcrypto-root/bin/openssl s_client -tlsextdebug -serverinfo 18 -connect localhost:4443
CONNECTED(00000003)
TLS server extension "unknown" (id=18), len=4
0000 - ff ff ff ff                                       ....
-----BEGIN SERVERINFO FOR EXTENSION 18-----
ABIABP////8=
-----END SERVERINFO FOR EXTENSION 18-----
```

The data format for the list of SCTs is described in the linked RFC above, and the following step by step instructions for creating an "sctl" file for haproxy would apply here: http://www.certificate-transparency.org/resources-for-site-owners/haproxy

colmmacc commented:
Wow, thank you for this! and especially for including test cases. I'm leaving a note here for @pzb to take a look at this too. He's our certificate transparency expert.

pzb commented:
This looks great.  SCTs provided via TLS extension is the gold standard as far as I'm concerned, so this is a great patch.

baldwinmatt commented:
This looks good - as we add support for more extensions, I think we should rethink the API here.  adding new functions to add new options is not sustainable.  We should instead offer an API that can take a list of options/extensions as an argument.

consider:

typedef struct {
  s2n_tls_extension_type type;
  uint8_t *data;
  uint32_t length;
} s2n_tls_extension;

we could then pass in an array of s2n_tls_extension* to the s2n_config_add_cert_chain_and_key function which could then be flattened onto the config struct.

aeijdenberg commented:
Matt - is that something that you'd like me to look at implementing as part of this patch?

baldwinmatt commented:
Not necessarily, rather thinking about what we should do going forward.

Sent from my iPhone

> On Jul 8, 2015, at 12:53, Adam Eijdenberg <notifications@github.com> wrote:
> 
> Matt - is that something that you'd like me to look at implementing as part of this patch?
> 
> —
> Reply to this email directly or view it on GitHub.
>

aeijdenberg commented:
I'd be happy to look at doing so, and moving the OCSP response to use it as well if you would find it helpful.  It did feel a bit clumsy to be adding 2 additional ways of specifying s2n_config_add_cert_chain_and_key.  It seems like it wouldn't be too hard to implement what you suggested.

aeijdenberg commented:
I changed the API surface as Matt suggested to think about.  Let me know what you think.

aeijdenberg commented:
Just cleaned up the changes into 2 commits, one for the CT support and the second for changing the API surface to take a list of s2n_tls_extensions.

Looks like travis has issues unrelated to this patch.

baldwinmatt commented:
looks good - the new API is certainly more flexible.

we are looking at ways to make this more extensible and maintainable in the future, esp. wrt adding multiple certs, but this is a step in the right direction.

aeijdenberg commented:
Cleaned up into a single commit. Travis appears to be giving flaky results a successful build.  Is that unique to this patch (indicating a problem with the patch) or are the kinks still getting straightened out with the the build/test process?

Please let me know if there's anything else I can do.

alexw91 commented:
I'm extremely sorry for delay. This has slipped through the cracks and hasn't been looked at for the last year and a half. 

Would you be willing to rebase this PR on top of master? 

We are now holding regular meetings to go through existing PR's, so I can guarantee that someone will take a look and review this PR if it is rebased.

aeijdenberg commented:
Hi @alexw91, just did a rebase now, fixed conflicts, updated tests etc. I've been out of the loop on CT myself for the past 15 months or so, so you may want to check with someone closer to the action as to whether this approach is still good.

pzb commented:
It would be great to get this added to s2n.  SCT delivery via TLS extension is the most flexible option for servers and allows SCT delivery with zero CA involvement.

alexw91 commented on tls/s2n_config.c:
```diff
@@ -427,9 +459,37 @@ int s2n_config_add_cert_chain_and_key_with_status(struct s2n_config *config, con
-        memcpy_check(config->cert_and_key_pairs->ocsp_status.data, status, length);
+    for (uint16_t ext_idx = 0; ext_idx < num_extensions; ext_idx++) {
+        switch (extensions[ext_idx].type) {
+            case S2N_EXTENSION_CERTIFICATE_TRANSPARENCY: {
```
Both of the cases in the switch statements can be simplified so as to avoid so much indentation. 

```
case S2N_EXTENSION_CERTIFICATE_TRANSPARENCY: {
    if (have_sct || extensions[ext_idx].data == NULL || extensions[ext_idx].length == 0) {
        S2N_ERROR(S2N_ERR_INVALID_SCT_LIST);
    } 
    have_sct = 1;
    GUARD(s2n_alloc(&config->cert_and_key_pairs->sct_list, extensions[ext_idx].length));
    memcpy_check(config->cert_and_key_pairs->sct_list.data, extensions[ext_idx].data, extensions[ext_idx].length);
} break;
```

aeijdenberg replied:
Fixed as suggested.

alexw91 commented on tls/s2n_connection.c:
```diff
@@ -637,3 +639,13 @@ int s2n_connection_is_managed_corked(const struct s2n_connection *s2n_connection
+
+const uint8_t *s2n_connection_get_sct_list(struct s2n_connection *conn, uint32_t *length)
+{
+    if (!length) {
```
I think our preferred way of doing this is to do `notnull_check(length);`

aeijdenberg replied:
The reason I did it this way was to match the method signature for the existing equivalent OCSP response method:

```c
extern const uint8_t *s2n_connection_get_ocsp_response(struct s2n_connection *conn, uint32_t *length);
```

I'm happy to go either way.

alexw91 commented on tls/s2n_client_extensions.c:
```diff
@@ -404,3 +417,9 @@ static int s2n_recv_client_renegotiation_info(struct s2n_connection *conn, struc
     return 0;
 }
+
+static int s2n_recv_client_sct_list(struct s2n_connection *conn, struct s2n_stuffer *extension)
```
`extension` parameter isn't used?

aeijdenberg replied:
Not sure if you need me to do anything here or not? I added a comment noting that we're skipping it.

alexw91 commented on tls/s2n_config.c:
```diff
@@ -447,7 +507,8 @@ int s2n_config_add_cert_chain_and_key_with_status(struct s2n_config *config, con
 int s2n_config_add_cert_chain_and_key(struct s2n_config *config, const char *cert_chain_pem, const char *private_key_pem)
 {
-    GUARD(s2n_config_add_cert_chain_and_key_with_status(config, cert_chain_pem, private_key_pem, NULL, 0));
+    GUARD(s2n_config_add_cert_chain_and_key_with_extensions(config, cert_chain_pem, private_key_pem,
```
Autoformat? Looks like this line was split at 100 chars, we usually aim for 120.

aeijdenberg replied:
Fixed.

alexw91 commented on tls/s2n_server_extensions.c:
```diff
@@ -163,3 +178,18 @@ int s2n_recv_server_status_request(struct s2n_connection *conn, struct s2n_stuff
+    sct_list.data = s2n_stuffer_raw_read(extension, sct_list.size);
+    notnull_check(sct_list.data);
+
+    GUARD(s2n_alloc(&conn->ct_response, sct_list.size));
```
You should be able to use `s2n_dup(&sct_list, &conn->ct_response);`

aeijdenberg replied:
Fixed.

aeijdenberg commented:
Thanks for the review @alexw91. I've added a new commit to address them.

Happy to squash them all back to a single commit when we're ready.
